### ISSUE TITLE
leastsq needlessly appends extra dimension for scalar problems

### ISSUE THREAD

**Issue by nschloe** [scipy/scipy#8532]
When treating a scalar problem, leastsq needlessly appends an extra dimension to the argument of the function to be minimized as well as the result. MCVE:
```python
from scipy.optimize import leastsq

def f(x):
    print(x.shape)
    return x


out = leastsq(f, 0.0)
print(out)
print()
leastsq(f, [0.0])
print(out)
```
Output:
```
(1,)
(1,)
(1,)
(1,)
(array([0.]), 4)

(1,)
(1,)
(1,)
(1,)
(array([0.]), 4)
```
Expected output:
```
()
()
()
()
(array(0.), 4)

(1,)
(1,)
(1,)
(1,)
(array([0.]), 4)
```
As a consequence of this, many codes have to select the "first" element in a scalar problem
```python
out, _ = leastsq(f, 0.0)
theta = out[0]  # this should be unnecessary
```
I personally ran into this when trying to clean up my `f(x)` which involves `dot`s and such, and wondered why things like
```python
array = something()

def f(theta):
    sc = numpy.array([numpy.sin(theta), numpy.cos(theta)])
    return numpy.dot(sc, array)
```
wouldn't work. (It was the superfluous dimension in `theta` that had the dot product fail.)

**Comment by sudheerachary** [comment]
I think this is due to the conversion of `x0` i.e (The starting estimate for the minimization) into a `numpy.array` in method `leastsq`, should this be handled if `x0` has only one value or add appropriate documentation for inputs of `func` in method [leastsq](https://docs.scipy.org/doc/scipy/reference/generated/scipy.optimize.leastsq.html)

**Comment by pv** [comment]
At this point, I would favor simply documentation-only fix as no doubt there is code out there relying on the current behavior, so it cannot just be changed.

The issue is that in general these routines work only on 1D vectors and don't preserve the input shape.

**Comment by nschloe** [comment]
Ah, you're right: It's the flatten() at https://github.com/scipy/scipy/blob/master/scipy/optimize/minpack.py#L377.

**Comment by sudheerachary** [comment]
Is this change fine for leastsq docstring or any additional changes required.  
```
diff --git a/scipy/optimize/minpack.py b/scipy/optimize/minpack.py
index a39bc15..2825ea4 100644
--- a/scipy/optimize/minpack.py
+++ b/scipy/optimize/minpack.py
@@ -275,7 +275,7 @@ def leastsq(func, x0, args=(), Dfun=None, full_output=0,
     Parameters
     ----------
     func : callable
-        should take at least one (possibly length N vector) argument and
+        should take at least one (possibly length N vector) argument of type numpy.array and
         returns M floating point numbers. It must not return NaNs or
         fitting might fail.
     x0 : ndarray
```

**Comment by andyfaff** [comment]
This is being addressed in #8666

**Comment by nschloe** [comment]
I've added a wrapper function for this in [scipyx](https://github.com/nschloe/scipyx#optimization), so I guess everyone's happy now. :smiley:
